cargo/jam/commands/pack.go

```diff
@@ -1,152 +1,104 @@
-type DirectoryDuplicator interface {
-	Duplicate(sourcePath, destPath string) error
+type packFlags struct {
+	buildpackTOMLPath string
+	output            string
+	version           string
+	offline           bool
+	stack             string
 }
 
-//go:generate faux --interface DependencyCacher --output fakes/dependency_cacher.go
-type DependencyCacher interface {
-	Cache(root string, dependencies []cargo.ConfigMetadataDependency) ([]cargo.ConfigMetadataDependency, error)
-}
-
-type Pack struct {
-	directoryDuplicator DirectoryDuplicator
-	configParser        ConfigParser
-	prePackager         PrePackager
-	dependencyCacher    DependencyCacher
-	tarBuilder          TarBuilder
-	fileBundler         FileBundler
-	stdout              io.Writer
-}
-
-func NewPack(
-	directoryDuplicator DirectoryDuplicator,
-	configParser ConfigParser,
-	prePackager PrePackager,
-	dependencyCacher DependencyCacher,
-	fileBundler FileBundler,
-	tarBuilder TarBuilder,
-	stdout io.Writer,
-) Pack {
-
-	return Pack{
-		directoryDuplicator: directoryDuplicator,
-		configParser:        configParser,
-		prePackager:         prePackager,
-		dependencyCacher:    dependencyCacher,
-		tarBuilder:          tarBuilder,
-		fileBundler:         fileBundler,
-		stdout:              stdout,
+func pack() *cobra.Command {
+	flags :=&packFlags{}
+	cmd := &cobra.Command{
+		Use:   "pack",
+		Short: "package buildpack",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return packRun(*flags)
+		},
 	}
-}
+	cmd.Flags().StringVar(&flags.buildpackTOMLPath, "buildpack", "", "path to buildpack.toml")
+	cmd.Flags().StringVar(&flags.output, "output", "", "path to location of output tarball")
+	cmd.Flags().StringVar(&flags.version, "version", "", "version of the buildpack")
+	cmd.Flags().BoolVar(&flags.offline, "offline", false, "enable offline caching of dependencies")
+	cmd.Flags().StringVar(&flags.stack, "stack", "", "restricts dependencies to given stack")
 
-func (p Pack) Execute(args []string) error {
-	var (
-		buildpackTOMLPath string
-		output            string
-		version           string
-		offline           bool
-		stack             string
-	)
-
-	fset := flag.NewFlagSet("pack", flag.ContinueOnError)
-	fset.StringVar(&buildpackTOMLPath, "buildpack", "", "path to buildpack.toml")
-	fset.StringVar(&output, "output", "", "path to location of output tarball")
-	fset.StringVar(&version, "version", "", "version of the buildpack")
-	fset.BoolVar(&offline, "offline", false, "enable offline caching of dependencies")
-	fset.StringVar(&stack, "stack", "", "restricts dependencies to given stack")
-	err := fset.Parse(args)
+	err := cmd.MarkFlagRequired("buildpack")
 	if err != nil {
-		return err
-	}
-
-	if buildpackTOMLPath == "" {
-		return errors.New("missing required flag --buildpack")
+		fmt.Fprintf(os.Stderr, "Unable to mark buildpack flag as required")
 	}
-
-	if output == "" {
-		return errors.New("missing required flag --output")
```

Comment:
Is output not a required flag anymore? When I run `jam-darwin pack --buildpack <path> --version 1.2.3` with no `--output` flag, rather than giving me a nice output it returns:
```
Error: failed to create output: failed to create tarball: open : no such file or directory
```
Maybe this should return the `required flag(s) "output" not set` too

Reply:
Good catch. We just need to add `cmd.MarkFlagRequired("output")`. I'm wondering why a unit test didn't catch this. Will look into.

Reply:
Added a few more test cases and marked output as required.